source/CCustomOpcodeSystem.cpp

```diff
@@ -399,16 +399,90 @@ namespace CLEO
 		return thread;
 	}
 
-	// read string parameter according to convention on strings. Always null terminated
-	char* ReadStringParam(CRunningScript *thread, char* buf, DWORD bufSize)
+	const char* ReadStringParam(CRunningScript *thread, char* buff, DWORD buffSize)
 	{
-		static char internal_buf[MAX_STR_LEN];
-		if (!buf) { buf = internal_buf; bufSize = MAX_STR_LEN; }
+		static char internal_buf[MAX_STR_LEN + 1]; // and terminator
 
-		if (bufSize > 0) buf[bufSize - 1] = '\0'; // buffer always terminated
-		if (bufSize <= 1) return buf; // no characters to read, done
+		// if available obtain pointer to the source, null terminated, string data
+		char* source = nullptr;
+		auto paramType = thread->PeekDataType();
+		auto arrayType = IsArray(paramType) ? thread->PeekArrayDataType() : eArrayDataType::ADT_NONE;
+		auto isVariableInt = IsVariable(paramType) && (arrayType == eArrayDataType::ADT_NONE || arrayType == eArrayDataType::ADT_INT);
+
+		if (IsImmInteger(paramType) || isVariableInt)
+		{
+			source = (char*)CLEO_PeekIntOpcodeParam(thread); // integer address to text buffer
+
+			if ((size_t)source <= CCustomOpcodeSystem::MinValidAddress)
+			{
+				CLEO_SkipOpcodeParams(thread, 1);
+				LOG_WARNING(thread, "Invalid '0x%X' pointer of input string argument #%d in script %s", opcodeParams[0].dwParam, CLEO_GetParamsHandledCount(), ScriptInfoStr(thread).c_str());

```

Comment:
replace opcodeParams[0].dwParam with source

Reply:
source is no longer here